### PR TITLE
feat: add Azure Monitor module

### LINKED ISSUE #630: Add Azure Monitor module
This issue is to track adding the [Azure Monitor](https://docs.microsoft.com/en-us/azure/azure-monitor/overview) module for terratest.

### PR DESCRIPTION
**Pull request by davidkpiano**
This PR adds the module for various Azure Monitor helpers.

Closes #630

### CODE REVIEW

File: modules/azure/monitor.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	client.Authorizer = *authorizer
+	return &client, nil
+}
+
+// ActionGroupExists indicates whether the speficied action group exists
+func ActionGroupExists(t testing.TestingT, actionGroupName string, resGroupName string, subscriptionID string) bool {
```
Comment (davesee) [inline modules/azure/monitor.go:111]: Add the error comment

File: modules/azure/monitor.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	client.Authorizer = *authorizer
+	return &client, nil
+}
+
+// GetVMInsightsE get diagnostics VM onboarding status
+func GetVMInsightsE(t testing.TestingT, resourceURI string, subscriptionID string) (*insights.VMInsightsOnboardingStatus, error) {
```
Comment (davesee) [inline modules/azure/monitor.go:79]: Should have a matching non-E func.

File: modules/azure/monitor.go (hunk 1 above)
Comment (davidkpiano) [reply modules/azure/monitor.go]: Removed the action group methods - I realized that they were redundant (there's an `actiongroup.go` file for this)

File: examples/azure/terraform-azure-monitor-example/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# Terraform Azure Monitor Example
+
+This folder contains a simple Terraform module that deploys resources in [Azure](https://azure.microsoft.com/) to demonstrate
+how you can use Terratest to write automated tests for your Azure Terraform code. This module deploys an Azure Storage Account and Azure Azure Key Vault with an Azure Monitor Diagnostic Setting.
+
+- A [Diagnostic Setting](https://docs.microsoft.com/en-us/azure/azure-monitor/platform/diagnostic-settings-template)
```
Comment (rguthriemsft) [inline examples/azure/terraform-azure-monitor-example/README.md:6]: remove culture from links

File: examples/azure/terraform-azure-monitor-example/main.tf [hunk 4]
```diff
@@ -0,0 +1,162 @@
+# ---------------------------------------------------------------------------------------------------------------------
+# DEPLOY AN AZURE MONITOR DIAGNOSTIC SETTING
+# This is an example of how to deploy an Azure Monitor Diagnostic Setting
+# for a key vault with a storage account.
+# ---------------------------------------------------------------------------------------------------------------------
+# See test/azure/terraform_azure_monitor_example_test.go for how to write automated tests for this code.
+# ---------------------------------------------------------------------------------------------------------------------
+
+provider "azurerm" {
+  version = "~> 2.20"
```
Comment (rguthriemsft) [inline examples/azure/terraform-azure-monitor-example/main.tf:10]: bump to 2.29

File: examples/azure/terraform-azure-monitor-example/main.tf [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+  name                        = "kv-${var.postfix}"
+  location                    = azurerm_resource_group.monitor.location
+  resource_group_name         = azurerm_resource_group.monitor.name
+  enabled_for_disk_encryption = true
+  tenant_id                   = data.azurerm_client_config.current.tenant_id
+  soft_delete_enabled         = true
+  # soft_delete_retention_days  = 7
+  purge_protection_enabled = false
+
+  sku_name = "standard"
```
Comment (rguthriemsft) [inline examples/azure/terraform-azure-monitor-example/main.tf:90]: is this the cheapest sku?  if so no problem, if not please adjust to cheapest.

File: modules/azure/monitor.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+func DiagnosticSettingsResourceExists(t testing.TestingT, diagnosticSettingsResourceName string, resourceURI string, subscriptionID string) bool {
+	exists, err := DiagnosticSettingsResourceExistsE(diagnosticSettingsResourceName, resourceURI, subscriptionID)
+	require.NoError(t, err)
+	return exists
+}
+
+// DiagnosticSettingsResourceExistsE indicates whether the diagnostic settings resource exists
+func DiagnosticSettingsResourceExistsE(diagnosticSettingsResourceName string, resourceURI string, subscriptionID string) (bool, error) {
+	_, err := GetDiagnosticsSettingsResourceE(diagnosticSettingsResourceName, resourceURI, subscriptionID)
+	if err != nil {
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:22]: follow pattern availability set uses for exist.   https://github.com/gruntwork-io/terratest/blob/48e24df893c925adef331c342b7d3979da08f51c/modules/azure/availabilityset.go#L21

File: modules/azure/monitor.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+		return nil, err
+	}
+
+	return &status, nil
+}
+
+// GetVMInsightsClientE gets a VM Insights client
+func GetVMInsightsClientE(t testing.TestingT, subscriptionID string) (*insights.VMInsightsClient, error) {
+	client := insights.NewVMInsightsClient(subscriptionID)
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:105]: need to lookup subscription similar to disgnostics setting client #L62

**Review by rguthriemsft — CHANGES_REQUESTED** [review]
@davidkpiano reviewed.  Please take a look at comments and resolve.

File: examples/azure/terraform-azure-monitor-example/main.tf (hunk 5 above)
Comment (davidkpiano) [reply examples/azure/terraform-azure-monitor-example/main.tf:90]: It is, yes.

File: test/azure/monitor_test.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
+// +build azure
+
+package test
```
Comment (rguthriemsft) [inline test/azure/monitor_test.go:3]: move unit test back to modules/azure

Comment (rguthriemsft) [inline test/azure/monitor_test.go:3]: once moved back to modules/azure rename package test to package azure.

File: modules/azure/monitor.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+// GetDiagnosticsSettingsResourceE gets the diagnostics settings for a specified resource
+func GetDiagnosticsSettingsResourceE(name string, resourceURI string, subscriptionID string) (*insights.DiagnosticSettingsResource, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := GetDiagnosticsSettingsClientE(subscriptionID)
+
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:48]: remove whitespace

File: modules/azure/monitor.go [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	settings, err := client.Get(context.Background(), resourceURI, name)
+
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:54]: remove whitespace

File: modules/azure/monitor.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+func GetDiagnosticsSettingsClientE(subscriptionID string) (*insights.DiagnosticSettingsClient, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	client := insights.NewDiagnosticSettingsClient(subscriptionID)
+	authorizer, err := NewAuthorizer()
+
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:72]: remove whitespace

File: modules/azure/monitor.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	// Get the Action Groups client
+	client := insights.NewActivityLogAlertsClient(subscriptionID)
+
+	// Create an authorizer
+	authorizer, err := NewAuthorizer()
+	if err != nil {
+		return nil, err
+	}
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:172]: add newline

File: modules/azure/monitor.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	client := insights.NewVMInsightsClient(subscriptionID)
+
+	authorizer, err := NewAuthorizer()
+
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:117]: remove whitespace.

File: modules/azure/monitor.go [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+// GetVMInsightsOnboardingStatusE get diagnostics VM onboarding status
+func GetVMInsightsOnboardingStatusE(t testing.TestingT, resourceURI string, subscriptionID string) (*insights.VMInsightsOnboardingStatus, error) {
+	client, err := GetVMInsightsClientE(t, subscriptionID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	status, err := client.GetOnboardingStatus(context.Background(), resourceURI)
+
```
Comment (rguthriemsft) [inline modules/azure/monitor.go:98]: remove whitespace

**Review by rguthriemsft — CHANGES_REQUESTED** [review]
@davidkpiano Please take a look at comments and resolve.

**Review by rguthriemsft — DISMISSED** [review]
Look good to me

**Comment by rguthriemsft** [comment]
Found an issue in CI, fixed and here is the passing run https://github.com/rguthriemsft/terratest/actions/runs/292015983

**Review by rguthriemsft — DISMISSED** [review]

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/315947114

**Review by rguthriemsft — APPROVED** [review]

File: examples/azure/terraform-azure-monitor-example/main.tf [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+# ---------------------------------------------------------------------------------------------------------------------
+
+resource "azurerm_key_vault" "monitor" {
+  name                        = "kv-${var.postfix}"
+  location                    = azurerm_resource_group.monitor.location
+  resource_group_name         = azurerm_resource_group.monitor.name
+  enabled_for_disk_encryption = true
+  tenant_id                   = data.azurerm_client_config.current.tenant_id
+  soft_delete_enabled         = true
+  # soft_delete_retention_days  = 7
```
Comment (yorinasub17) [inline examples/azure/terraform-azure-monitor-example/main.tf:85]: NIT: Remove dead code

**Review by yorinasub17 — APPROVED** [review]
LGTM! I found one nit.

File: examples/azure/terraform-azure-monitor-example/main.tf (hunk 15 above)
Comment (davidkpiano) [reply examples/azure/terraform-azure-monitor-example/main.tf:85]:
```suggestion
```

**Comment by yorinasub17** [comment]
Update LGTM! However, it looks like we need to run `terraform fmt` on the files. I think we can merge this in once we run that through!

**Comment by rguthriemsft** [comment]
@yorinasub17 I ran terraform fmt on the files in the example for this PR and picked up one issue which I just pushed commit for.  Do you need me to run the CI again for this or are we good?

**Comment by rguthriemsft** [comment]
[Microsoft CI Bot] TL;DR; success :thumbsup:

You can check the status of the CI Pipeline logs here ; https://github.com/rguthriemsft/terratest/actions/runs/392824989

**Review by yorinasub17 — APPROVED** [review]
Updates LGTM and confirmed our precommit stage passes, so this is good to go! Will merge and release shortly.

**Review by zackproser — APPROVED** [review]
LGTM

**Comment by yorinasub17** [comment]
Thanks Zack! Merging and releasing this now!

**Pull request merged**